### PR TITLE
Checkpoint 3 Review

### PR DESCRIPTION
**Pull request by femioladeji**
The APIs can be tested by using POSTMAN to make requests to https://dmsystem-api.herokuapp.com/

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    "mocha": "^3.1.2",
+    "supertest": "^2.0.1"
+  },
+  "dependencies": {
+    "bcrypt": "^0.8.7",
+    "body-parser": "^1.15.2",
+    "chai": "^3.5.0",
+    "dotenv": "^2.0.0",
+    "express": "^4.14.0",
+    "faker": "^3.1.0",
```
Comment (sslgeorge) [inline package.json]: Dev dependency .

File: server/controllers/authentication.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+              done: false,
+              message: 'Token authentication failed' });
+          }
+          req.decoded = decoded;
+          req.token = req.headers.authorization;
+          next();
+        });
+    } else {
+      return res.status(401).send({ done: false, message: 'Please login' });
+    }
```
Comment (sslgeorge) [inline server/controllers/authentication.js]: Not much of a big deal, but normally following a defensive approach is much more easier to read, you will first check if the headers do not contain the authorization token and return early if not, that way you just have an if block without necessarily the else part and an easier to ready code.

File: server/controllers/users-controller.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+   */
+  static login(req, res) {
+    db.user.findAll({
+      where: {
+        username: req.body.username
+      },
+      include: [db.role]
+    }).then((userDetails) => {
+      if (userDetails[0]) {
+        const userInfo = UsersController.fetchUser(userDetails[0].dataValues);
```
Comment (sslgeorge) [inline server/controllers/users-controller.js:73]: Does `userDetails` not contain the data in `userInfo`?

File: server/controllers/users-controller.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+          });
+        }
+        return res.status(401).send({
+          done: false,
+          message: 'Invalid password'
+        });
+      }
+      return res.status(401).send({
+        done: false,
+        message: 'Invalid username'
```
Comment (sslgeorge) [inline server/controllers/users-controller.js]: During login you don't want to tell the users the exact field they have wrongly inputed. Just send a general message saying one of the field is invalid.

File: server/controllers/users-controller.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  /**
+   * method getUserDocuments to get a user documents
+   * @param {object} req - request details
+   * @param {object} res - response details
+   * @return {array} array of documents;
+   */
+  static getUserDocuments(req, res) {
+    const userId = parseInt(req.params.id, 10);
+    db.document.findAll({ ownerId: userId, order: [['createdAt', 'DESC']] })
+      .then(documents => res.status(200).send({ doc: documents }));
```
Comment (sslgeorge) [inline server/controllers/users-controller.js]: using `res.send` means the application will not be able to convert non JSON data, like null and undefined. You might want to consider using `res.json` instead

File: server/models/user.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        user.hasMany(models.document, {
+          foreignKey: {
+            name: 'ownerId',
+            allowNull: false,
+          },
+          onDelete: 'CASCADE'
+        });
+      }
+    },
+    hooks: {
```
Comment (sslgeorge) [inline server/models/user.js:48]: Nice!

File: server/controllers/users-controller.js (hunk 3 above)
Comment (femioladeji) [reply server/controllers/users-controller.js:73]: No, userInfo is just like a formatted userDetails. The `fetchUser` functions saves me the extra stress of writing userDetails[0].dataValues to get each property of the object

Comment (sslgeorge) [reply server/controllers/users-controller.js:73]: Okay, I think that makes sense, good work.

**Pull request merged**